L67-cpp-callback-api.md

```diff
@@ -0,0 +1,125 @@
+
+### Related proposals
+
+* The C++ callback API is built on top of a new [callback completion queue in core](https://github.com/grpc/proposal/pull/181)
+* The API structure has substantial similarities to the gRPC-Node and gRPC-Java APIs
+
+## Proposal
+
+The callback API is designed to have the performance and thread scalability of an asynchronous API without the burdensome programming model of the completion-queue-based model. In particular, the following are fundamental guiding principles of the API:
+
+* Library directly calls user-specified code at the completion of RPC actions.
+* No explicit polling required for notification of completion of RPC actions.
+   - In practice, these requirements mean that there must be a library-controlled poller for monitoring such actions.
+* As in the synchronous API, server RPCs have an application-defined method handler function as part of their service definition. The library invokes this method handler when a new server RPC starts.
+* Like the synchronous API and unlike the completion-queue-based asynchronous API, there is no need for the application to "request" new server RPCs. Server RPC context structures will be allocated and have their resources allocated as and when RPCs arrive at the server.
+
+### Reactor model
+
+The most general form of the callback API is built around a _reactor_ model. Each type of RPC has a reactor base class provided by the library. These types are:
+
+* `ClientUnaryReactor` for unary calls
+* `ClientBidiReactor` for bidi-streaming 
+* `ClientReadReactor` for server-streaming 
+* `ClientWriteReactor` for client-streaming 
+* `ServerUnaryReactor` for unary calls
+* `ServerBidiReactor` for bidi-streaming 
+* `ServerReadReactor` for client-streaming 
+* `ServerWriteReactor` for server-streaming 
+
+Client RPC invocations from a stub provide a reactor pointer as one of their arguments, and the method handler of a server RPC must return a reactor pointer.
+
+These base classes provide three types of methods:
+
+1. Operation-initiation methods: start an asynchronous activity in the RPC. These are methods provided by the class and are not virtual.
```

Comment:
I suggest reorganizing this such that each method is described in its own bullet.  For example:

- `StartCall()`: (Client only.)  Initiates a new call with a specific set of initial metadata.  Must be called exactly once.  No reads or writes will actually be started until this is called (i.e., any previous calls to `StartRead()` or `StartWrite()` will be queued until `StartCall()` is invoked).
- `StartSendInitialMetadata()`: (Server only.) Sends initial metadata.  To be used in cases where initial metadata should be sent without sending a message.  Optional; if not called, initial metadata will be sent when `StartWrite()` or `Finish()` is called.  May not be invoked more than once.
- `StartRead()`: Starts a read of a message.  `OnReadDone()` will be invoked when the read is complete.  Only one read may be outstanding at any given time.  On the client, if invoked before `StartCall()`, 
- `StartWrite()`: Starts a write of a message.  `OnWriteDone()` will be invoked when the write is complete.  Only one write may be outstanding at any given time.
- `StartWritesDone()`: Similar to `StartWrite()`, but indicates that this will be the last message sent on the call.  When the operation is complete, `OnWritesDoneDone()` will be invoked.  May be used at most once per call.
- `Finish()`:  (Server only.) Sends completion status to the client.  Must be invoked exactly once, even for calls that have already been cancelled.

Similar thing for the completion methods and the completion-prevention methods.

Reply:
Sounds good, will do.

Reply:
Thanks for the great suggestion. Done.